File: devtools/shared/fronts/device.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

const { Cu } = require("chrome");
const { deviceSpec } = require("devtools/shared/specs/device");
const {
  FrontClassWithSpec,
  registerFront,
} = require("devtools/shared/protocol");
const defer = require("devtools/shared/defer");

class DeviceFront extends FrontClassWithSpec(deviceSpec) {
  constructor(client) {
    super(client);

    // Attribute name from which to retrieve the actorID out of the target actor's form
    this.formAttributeName = "deviceActor";
  }

  screenshotToBlob() {
    return this.screenshotToDataURL().then(longstr => {
      return longstr.string().then(dataURL => {
        const deferred = defer();
        longstr.release().catch(Cu.reportError);
        const req = new XMLHttpRequest();
        req.open("GET", dataURL, true);
        req.responseType = "blob";
        req.onload = () => {
          deferred.resolve(req.response);
        };
        req.onerror = () => {
          deferred.reject(req.status);
        };
        req.send();
        return deferred.promise;
      });
    });
  }
}

exports.DeviceFront = DeviceFront;
registerFront(DeviceFront);

File: dom/webauthn/u2f-hid-rs/src/stub/device.rs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

use std::io;
use std::io::{Read, Write};
use u2ftypes::U2FDevice;

pub struct Device {}

impl Device {
    pub fn new(path: String) -> io::Result<Self> {
        panic!("not implemented");
    }

    pub fn is_u2f(&self) -> bool {
        panic!("not implemented");
    }
}

impl Read for Device {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        panic!("not implemented");
    }
}

impl Write for Device {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        panic!("not implemented");
    }

    fn flush(&mut self) -> io::Result<()> {
        panic!("not implemented");
    }
}

impl U2FDevice for Device {
    fn get_cid<'a>(&'a self) -> &'a [u8; 4] {
        panic!("not implemented");
    }

    fn set_cid(&mut self, cid: [u8; 4]) {
        panic!("not implemented");
    }
}
